### ISSUE TITLE
jax2tf polymorphic shapes don't work with rng keys

### ISSUE THREAD

**Issue by sygi** [google/jax#7398]
While related to my previous issue (#7344), this one manifests itself in tf2.

```python
import tensorflow.compat.v2 as tf_v2
tf_v2.enable_v2_behavior()
import jax
import functools
from jax.experimental import jax2tf

def fn(key, x):
  return jax.random.categorical(key, x)

k = jax.random.PRNGKey(42)
part = functools.partial(fn, k)

h = tf_v2.function(jax2tf.convert(part, polymorphic_shapes=["(b, 5)"], with_gradient=False), autograph=False)
h3 = tf_v2.function(jax2tf.convert(fn, polymorphic_shapes=["(2,)", "(b, 5)"], with_gradient=False), autograph=False)
concrete = h.get_concrete_function(tf_v2.TensorSpec([None, 5], tf_v2.float32))
concrete2 = h3.get_concrete_function(tf_v2.TensorSpec([2], tf_v2.uint32), tf_v2.TensorSpec([None, 5], tf_v2.float32))
```

Error:
```
google3/third_party/py/jax/_src/random.py in _random_bits(key, bit_width, shape)
    302     key = fold_in(key, axis_index)
    303   size = prod(shape.positional)
--> 304   max_count = int(np.ceil(bit_width * size / 32))
    305 
    306   nblocks, rem = divmod(max_count, jnp.iinfo(np.uint32).max)

TypeError: unsupported operand type(s) for /: '_DimPolynomial' and 'int'
```
(`size` is `b*5`, which multiplies with `bit_width` but fails to divide by 32).

Trying to pass the key batched doesn't help, as it's assumed that the key is always `(2,)`:
```python
h2 = tf_v2.function(jax2tf.convert(fn2, polymorphic_shapes=["(b, 2)", "(b, 5)"], with_gradient=False), autograph=False)
concrete2 = h2.get_concrete_function(tf_v2.TensorSpec([None, 2], tf_v2.float32), tf_v2.TensorSpec([None, 5], tf_v2.float32))
```
gives:
```
InconclusiveDimensionOperation: Dimension polynomial comparison 'b' == '2' is inconclusive

This error arises for arithmetic or comparison operations with shapes that
are non-constant, and the result of the operation cannot be represented as
a polynomial of dimension variables, or a boolean constant (for comparisons).
```

**Comment by gnecula** [comment]
I can reproduce this, there are two bugs. One is to handle "b*5 * bit_width / 32" for DimPolynomial. I will fix this first. 

Unfortunately, it seems that this will not be enough to get you unstuck, because the PRNG does all sorts of computations with shapes that are not yet compatible with shape polymorphism. I'll need to think some more how to handle this.

**Comment by gnecula** [comment]
This is partially fixed by #7406. However, it will only work when the size of the desired shape is even for all values of the shape variables. This in practice means that the product of the known dimensions should be even. 

The reason is that the jax.random internals do `if size % 2 == 0` in the process of computing _random_bits. If size is a polynomial that is not obviously even, e.g., `5*b`, then the only safe thing for the shape polymorphism machinery is to throw an error.
However, if the polynomial is `6*b` then it can tell it is even for all values of `b`, so it is safe to take one branch.

A better solution is to rewrite the jax.random to be safe for shape polymorphism.

**Comment by gnecula** [comment]
I am going to close this for now, please feel free to reopen if you find other issued with shape polymorphism and lax.random.
